src/modm/driver/encoder/encoder_impl.hpp

```diff
@@ -0,0 +1,44 @@
+#ifndef MODM_ENCODER_HPP
+#error "Don't include this file directly, use 'encoder.hpp' instead!"
+#endif
+
+template<typename PH_A, typename PH_B, modm::encoder::Division_t DIVISION, std::signed_integral UnderlyingType>
+inline void
+modm::Encoder<PH_A, PH_B, DIVISION, UnderlyingType>::update()
+{
+	const auto read = Phases::read();
+	
+	UnderlyingType raw = 0;
+	if (read & 0b10) raw = 3;
+	if (read & 0b01) raw ^= 1;  // convert gray to binary
+
+	const auto diff = raw_last - raw;
+	if (diff & 1)
+	{
+		raw_last = raw;               // store new as next last
+		enc_delta += (diff & 2) - 1;  // bit 1 = direction (+/-)
+	}
+}
+
+template<typename PH_A, typename PH_B, modm::encoder::Division_t DIVISION, std::signed_integral UnderlyingType>
+UnderlyingType
+modm::Encoder<PH_A, PH_B, DIVISION, UnderlyingType>::getValue()
+{
+	::modm::atomic::Lock lock;
+	UnderlyingType val = enc_delta;  // counts since last call
+	switch (DIVISION)
+	{
+		case encoder::Division::ONE:
+			enc_delta = 0;
+			break;
+		case encoder::Division::TWO:
+			enc_delta = val & 1;
+			val >>= 1;
+			break;
+		case encoder::Division::FOUR:
+			enc_delta = val & 3;
+			val >>= 2;
+	}
+	::modm::atomic::Unlock unlock;
```

Comment:
No, just remove this line, you don't need it. 

The `modm::atomic::Lock` is a class with a constructor, which calls disables IRQs on construction, ie. line 27, and restores IRQs on destruction when the object leaves scope, here on line 43. [Here's the implementation for AVR](https://github.com/modm-io/modm/blob/develop/src/modm/platform/core/avr/atomic_lock_impl.hpp#L30-L48). Instead of calling `sei()` in the destructor, it restores SREG, so that it doesn't re-enable interrupts that have previously been (intentionally) disabled for example in nested atomic locks.

This also works if you have multiple returns or throw exceptions in your function. [This idiom is known as RAII in C++](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization).

Reply:
Now i understood the mechanism.